### PR TITLE
Php81compat

### PR DESCRIPTION
**Pull request by HamishSlater**
Refactor 'pass by reference' use

### CODE REVIEW

File: src/Globals.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace SMW;
+
+class Globals {
+
+	/**
+	 * PHP8.1 Compatibility
+	 * @since 4.0
```
Comment (JeroenDeDauw) [inline src/Globals.php:9]: Not since 4.0. Just drop that tag and perhaps mark the class as @internal

File: src/GroupPermissions.php [hunk 2]
```diff
@@ -44,15 +47,19 @@ public function initPermissions( &$vars ) {
...
 			if ( !isset( $vars['wgGroupPermissions'][$group] ) ) {
-				$vars['wgGroupPermissions'][$group] = [];
+				$newVars['wgGroupPermissions'][$group] = [];
 			}
 
-			$vars['wgGroupPermissions'][$group] = array_merge( $rights, $vars['wgGroupPermissions'][$group] );
+			$newVars['wgGroupPermissions'][$group] = array_merge( $rights, $vars['wgGroupPermissions'][$group] );
 		}
+
+		Globals::replace( $newVars );
```
Comment (JeroenDeDauw) [inline src/GroupPermissions.php:60]: Can't the caller do this side effect? Weird to both return the new value and modify globals. Like this the test also does not check the production behavior.

File: src/NamespaceManager.php [hunk 3]
```diff
@@ -60,7 +60,7 @@ public function init( $vars ) {
 		$newVars = [];
 
 		if ( !$this->isDefinedConstant( 'SMW_NS_PROPERTY' ) ) {
-			Globals::replace( $newVars = $this->initCustomNamespace( $vars ) );
+			Globals::replace( $newVars = $this->initCustomNamespace( $vars )['newVars'] );
```
Comment (JeroenDeDauw) [inline src/NamespaceManager.php:63]: Why is initCustomNamespace returning two variables in an array?

Comment (HamishSlater) [reply src/NamespaceManager.php:63]: initCustomNamespace() originally returned $instance.
For testing it also needs to return $newVars.

So a quick solution was to return:
	'instance' => $instance,
	'newVars' => $newVars

I considered setting a class variable '$instance', then call it from 'initCanonicalNamespaces()';
however, both methods are static.

**Comment by JeroenDeDauw** [comment]
Merge conflicts with master

@HamishSlater I'm unfollowing this PR. Please nudge me when it is ready for review

**Comment by HamishSlater** [comment]
> Merge conflicts with master
> 
> @HamishSlater I'm unfollowing this PR. Please nudge me when it is ready for review

@JeroenDeDauw Github states that there are no conflicts.
I have pulled the branch to include the recent commits.
It is awaiting review.

**Comment by JeroenDeDauw** [comment]
image

**Comment by HamishSlater** [comment]
image

**Comment by JeroenDeDauw** [comment]
Wrong upstream?

**Comment by malberts** [comment]
There is some underlying issue with this branch diverging from upstream master which requires a rebase outside of GitHub.

Perhaps it will be easier to recreate the branch with your changes. I tested this in #5353 and it is not complaining about a conflict.

**Pull request closed without merging**